scarb/src/bin/scarb/commands/commands.rs

```diff
@@ -1,9 +1,139 @@
+struct CommandsList {
+    commands: BTreeMap<String, CommandInfo>,
+}
+
+impl Message for CommandsList {
+    fn text(self) -> String {
+        let mut text = String::from("Installed Commands:\n");
+        for (name, info) in self.commands.iter() {
+            text.push_str(&format!("{:<20}: {}\n", name, info));
+        }
+        text
+    }
+
+    fn structured<S: Serializer>(self, ser: S) -> Result<S::Ok, S::Error> {
+        self.commands.serialize(ser)
+    }
+}
+
+fn list_commands(config: &Config, builtins: &mut BTreeMap<String, Option<String>>) -> CommandsList {
+    let prefix = EXTERNAL_CMD_PREFIX;
+    let suffix = env::consts::EXE_SUFFIX;
+
+    let mut commands = BTreeMap::new();
+    for dir in config.dirs().path_dirs.iter() {
+        let Ok(entries) = fs::read_dir(dir) else { continue; };
+        for entry in entries.filter_map(|e| e.ok()) {
+            let path = entry.path();
+            let Some(filename) =  path.file_name().and_then(|s| s.to_str()) else{ continue; };
+            if !filename.starts_with(prefix) || !filename.ends_with(suffix) {
+                continue;
+            }
+            if is_executable(entry.path()) {
+                let end = filename.len() - suffix.len();
+                commands.insert(
+                    filename[prefix.len()..end].to_string(),
+                    CommandInfo::External { path: path.clone() },
+                );
+            }
+        }
+    }
+
+    // In case of name conflict, builtin commands take precedence.
+    let mut builtin_commands = BTreeMap::new();
+    for (name, about) in builtins.iter() {
+        builtin_commands.insert(
+            name.clone(),
+            CommandInfo::BuiltIn {
+                about: about.clone(),
+            },
+        );
+    }
+    commands.append(&mut builtin_commands);
+
+    CommandsList { commands }
+}
 
 #[tracing::instrument(skip_all, level = "info")]
 pub fn run(config: &Config) -> Result<()> {
-    config.ui().print("Installed commands:");
-    todo!("not implemented yet.")
+    let mut builtins = ScarbArgs::get_builtin_subcommands();
+    config.ui().print(list_commands(config, &mut builtins));
+    Ok(())
+}
+
+#[cfg(test)]
+#[cfg(unix)]
+mod tests {
+    use super::{list_commands, CommandInfo};
+    use assert_fs::prelude::*;
+    use assert_fs::TempDir;
+    use camino::Utf8Path;
+    use scarb::core::Config;
+    use std::collections::BTreeMap;
+    use std::fs::set_permissions;
+    use std::os::unix::fs::PermissionsExt;
+    use std::path::PathBuf;
+
+    #[test]
+    fn cmd_list() {
+        let t = TempDir::new().unwrap();
+
+        let cache_dir = TempDir::new().unwrap();
+        let config_dir = TempDir::new().unwrap();
+        let manifest = t.child("Scarb.toml");
+        let path_dir = t.child("bin");
+        let sub_cmd = path_dir.child("scarb-hello");
+        sub_cmd.touch().unwrap();
+        set_permissions(sub_cmd.path(), std::fs::Permissions::from_mode(0o755)).unwrap();
```

Comment:
You can move `make_executable` to `scarb::process` and use it here and thus this test will work on Windows